### ISSUE TITLE
[Question] Netdata nginx configuration additional param possibility?

### ISSUE THREAD

**Issue by jeremyjpj0916** [firehol/netdata#3395]
Your nginx config plugin:
https://github.com/firehol/netdata/blob/master/conf.d/python.d/nginx.conf

I have a case where I have two containers starting up in the same server same time on the cloud and I would like to set a hard delay before netdata tries to hit the nginx stub_status to start getting results. Currently netdata tries to hit it as soon as it can and seems to stop trying even if I set retries up to a massive number, I would like a delay before the code that calls out to nginx executes:

```
update_every : 10
priority     : 90100
delay        : 3600 <---- seconds (new potential arg for delaying the initial call to nginx)

local:
  url     : 'http://localhost/stub_status'
  retries : 10
```

But seeing as this may not help everyone you may not feel it should be a feature, which is fine by me but if you could direct me where in the actual source code files I can go and what 1 liner "sleep" I can add in this call to wait a while before starting up that would work wonders for me.

Other than this little issue and then the statsD graph one I know you are aware of then I think Netdata will be ready for my production use case!

Thanks,
Jeremy

**Comment by l2isbad** [comment]
> Currently netdata tries to hit it as soon as it can and seems to stop trying even if I set retries up to a massive number

https://github.com/firehol/netdata/blob/master/conf.d/python.d/nginx.conf#L48-L52

**Comment by jeremyjpj0916** [comment]
@l2isbad You built the nginx/nginx plus plugins right? Thanks for pointing that out to me, I did not catch that arg! You thought of that scenario 👍 Closing issue, thanks!

On a side note I assume you might be a nginx engineer? If so could you entertain me a look over of this conf file to see if I am doing proper optimization options within nginx? This was my first real attempt at using nginx as a reverse proxy and I just want to make sure I cover all bases, generally payloads will be 8kb or so, but we wanna support up to 10mb request / 20mb responses max too, do you have any suggestions or improvements based on what I have?:

```
charset UTF-8;

error_log /dev/stderr notice;

#Optimizations?
tcp_nodelay on;
tcp_nopush on;
proxy_buffer_size 128k;
proxy_buffers 20 256k;
proxy_busy_buffers_size 256k;
reset_timedout_connection on;
keepalive_requests 200;
keepalive_timeout 75s;

client_max_body_size 0;
proxy_ssl_server_name on;
underscores_in_headers on;

lua_package_path './?.lua;./?/init.lua;;;';
lua_package_cpath ';;';
lua_socket_pool_size 30;
lua_max_running_timers 4096;
lua_max_pending_timers 16384;
lua_shared_dict kong                5m;
lua_shared_dict kong_cache          256m;
lua_shared_dict kong_process_events 5m;
lua_shared_dict kong_cluster_events 5m;
lua_shared_dict kong_healthchecks   5m;
lua_shared_dict kong_cassandra      5m;
lua_socket_log_errors off;
lua_ssl_trusted_certificate '/usr/local/kong/ssl/kongcert.pem';
lua_ssl_verify_depth 2;

init_by_lua_block {
    kong = require 'kong'
    kong.init()
}

init_worker_by_lua_block {
    kong.init_worker()
}

proxy_next_upstream_tries 999;

upstream kong_upstream {
    server 0.0.0.1;
    balancer_by_lua_block {
        kong.balancer()
    }
    keepalive 100;
}

server {
    server_name kong;
    listen 0.0.0.0:8000;
    error_page 400 404 408 411 412 413 414 417 /kong_error_handler;
    client_header_buffer_size 8k;
    large_client_header_buffers 100 32k;
    error_page 500 502 503 504 /kong_error_handler;

    access_log off;
    error_log /dev/stderr notice;

    client_body_buffer_size 5m;

    listen 0.0.0.0:8443 ssl http2;
    ssl_certificate /usr/local/kong/ssl/kongcert.crt;
    ssl_certificate_key /usr/local/kong/ssl/kongprivatekey.key;
    ssl_protocols TLSv1.2;
    ssl_certificate_by_lua_block {
        kong.ssl_certificate()
    }

    ssl_session_cache shared:SSL:10m;
    ssl_session_timeout 10m;
    ssl_prefer_server_ciphers on;
    ssl_ciphers ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256;


    real_ip_header     X-Real-IP;
    real_ip_recursive  off;

    location / {
        set $upstream_host               '';
        set $upstream_upgrade            '';
        set $upstream_connection         '';
        set $upstream_scheme             '';
        set $upstream_uri                '';
        set $upstream_x_forwarded_for    '';
        set $upstream_x_forwarded_proto  '';
        set $upstream_x_forwarded_host   '';
        set $upstream_x_forwarded_port   '';

        rewrite_by_lua_block {
            kong.rewrite()
        }

        access_by_lua_block {
            kong.access()
        }

        proxy_http_version 1.1;
        proxy_set_header   Host              $upstream_host;
        proxy_set_header   Upgrade           $upstream_upgrade;
        proxy_set_header   Connection        $upstream_connection;
        proxy_set_header   X-Forwarded-For   $upstream_x_forwarded_for;
        proxy_set_header   X-Forwarded-Proto $upstream_x_forwarded_proto;
        proxy_set_header   X-Forwarded-Host  $upstream_x_forwarded_host;
        proxy_set_header   X-Forwarded-Port  $upstream_x_forwarded_port;
        proxy_set_header   X-Real-IP         $remote_addr;
        proxy_pass_header  Server;
        proxy_pass_header  Date;
        proxy_ssl_name     $upstream_host;
        proxy_pass         $upstream_scheme://kong_upstream$upstream_uri;

        header_filter_by_lua_block {
            kong.header_filter()
        }

        body_filter_by_lua_block {
            kong.body_filter()
        }

        log_by_lua_block {
            kong.log()
        }
    }

    location = /kong_error_handler {
        internal;
        content_by_lua_block {
            kong.handle_error()
	}
    }
}

server {
    server_name kong_admin;
    listen 0.0.0.0:8001;

    access_log off;
    error_log /dev/stderr notice;

    client_max_body_size 10m;
    client_body_buffer_size 10m;

    listen 0.0.0.0:8444 ssl;
    ssl_certificate /usr/local/kong/ssl/admin-kong-default.crt;
    ssl_certificate_key /usr/local/kong/ssl/admin-kong-default.key;
    ssl_protocols TLSv1.2;

    ssl_session_cache shared:SSL:10m;
    ssl_session_timeout 10m;
    ssl_prefer_server_ciphers on;
    ssl_ciphers ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256;

    location / {
        default_type application/json;
        content_by_lua_block {
            kong.serve_admin_api()
        }
    }

    location /nginx_status {
        access_log off;
        stub_status;
    }

    location /robots.txt {
        return 200 'User-agent: *\nDisallow: /';
    }
}
```

I may have to set my buffers even higher I believe to support our largest payload goals. 

-Jeremy

**Comment by l2isbad** [comment]
@jeremyjpj0916 
Sorry, i can't help you there :cry: I am a network guy irl.

**Comment by l2isbad** [comment]
the only advice is make sure you are using `web_log` plugin to chart `nginx` logs 😄

**Comment by jeremyjpj0916** [comment]
@l2isbad No problem man you have done a great deal helping me out already :) ! I do have a question on the web_log plugin if you developed that out(Sorry last question :P ):

```
nginx_log:
  name  : 'nginx_log'
  path  : '/var/log/nginx/access.log'
```

Say that I am pointing my nginx not to any direct access.log file or error.log, but rather pointing to things such as /dev/stdout /dev/stderr (docker log collector standard output/err), can this plugin tail the stdout logs? And can it decipher which of the logs are nginx access logs vs say something I randomly decided to print out to stdout? It does seem like a really neat feature, I just don't log to an actual file since OpenShift Enterprise manages my docker logs in the bg. Note my specific config of nginx and netdata are two separate containers running  on the same Pod if you are familiarized with Kubernetes/OpenShift.

Thanks!
-Jeremy

**Comment by l2isbad** [comment]
> can this plugin tail the stdout logs

Unfortunately no

**Comment by jeremyjpj0916** [comment]
@l2isbad gotcha, I suppose I could add logrotate and let it log to files again but its low priority, will just pass on the web_log for now. Maybe I will peruse the source code and see if I can add a way for it to capture it from the stdout :) . Thanks again for your responsiveness!

**Comment by Ferroin** [comment]
@jeremyjpj0916, In response to your question about Nginx optimizations:

DISCLAIMER: I am not an Nginx expert, just a guy who runs a couple of Nginx-based web servers and is somewhat obsessive about performance.

There are a couple of things I can think of that you might change to improve performance:

1. Enable multithreading.  Serving web pages is I/O bound (either disk or network), so it's actually a case where you can boost performance with multi-threading even if you're running on a single processor.  To get any real benefit though, it has to be used with AIO (and thus DirectIO).  The general approach involves adding a line like the following outside the `http` element:

        thread_pool pool0 threads=4 max_queue=65536;

    And then putting the following few lines inside the `http` element:

        aio threads=pool0;
        directio 4m;

    The `thread_pool` line defines a pool of threads that can be referenced later in the configuration, in chis case called `pool0`, with a maximal queue-size of 65536 elements.  The `aio` directive specifies how to perform asynchronous IO, in this case using the thread pool defined earlier to do it, and the `directio` statement specifies the minimal response size that will be processed with DirectIO (and therefore handled by the AIO interface).

2. Enable multi-processing if you reasonably can.  This will also help with parallelization, but in a different way, as each worker process can handle a separate request.  The config option for this is `worker_processes`, either set it to a positive number to use that many processes, or set it to `auto` to use as many worker processes as you have logical CPU's.  If you are really picky, you can also uses `worker_cpu_affinity` to bind each worker process to a given CPU.

3. Look at some fine tuning for your socket options in the `listen` directive.  In particular, if you see lots of individual connections, you might want to look at the `deferred` option, it can save som processing time on new connections.  If you end up enabling multiprocessing, you probably also want the `reuseport` option, which pushes the load-balancing of new connections out to the kernel instead of having Nginx do it itself (which saves some more processing power).

4. You're enabling HTTP/2 and configuring regular keepalive settings, but you're not doing any keepalive related configuration for HTTP/2.  There are actually special config options specifically for HTTP/2 for this.  `http2_max_requests` is equivalent to `keepalive_requests`, and `http2_idle_timeout` is equivalent to `keepalive_timeout` (but without the ability to advertise a shorter timeout).

Your other optimizations appear to be reasonable ones, though you might consider tweaking some of the buffer values a bit to see how they impact performance.  Most of the other advice I would have is focused more around serving static files (stuff like using `sendfile` and `open_file_cache`), but you appear to be serving purely dynamically generated content here other than `robots.txt`, so none of that is likely to be much help.

In general though, I very much agree with @l2isbad, look into getting the `web_log` plugin set up, and more specifically check the documentation for it for info on how to get timing statistics out of Nginx.  Even if you just use it for temporary testing of the configuration, it's one of the best options out there for figuring out how your web server is performing.

**Comment by jeremyjpj0916** [comment]
@Ferroin Wow thanks for your detailed response. I am blown away by how nice the open source community is. I will look into all these configuration options you have mentioned to me, especially good catch on the http2 extra options I missed! For part 2 I do indeed have that auto config set, I suppose I actually forgot to include my higher level .conf file that the above conf feeds into:

```
worker_processes auto;                                                          
daemon off;                                                                     
                                                                                
env OPTUM_SPLUNK_HOST;                                                          
env KONG_SSL_CERT_KEY;                                                          
pid pids/nginx.pid;                                                             
error_log /dev/stderr notice;                                                   
                                                                                
worker_rlimit_nofile 1048576;                                                   
                                                                                
events {                                                                        
    worker_connections 16384;                                                   
    multi_accept on;                                                            
}                                                                               
                                                                                
http {                                                                          
    include 'nginx-kong.conf';              <---- this is the above conf I included first                                    
}    
```

So that gives you more of a full picture. But thanks again for your input, I really do appreciate it. And you sound pretty close to an expert :) , don't sell yourself short of it hah.
